File: src/main/java/io/github/gaming32/superpack/ProgressDialog.java
package io.github.gaming32.superpack;

import com.jthemedetecor.OsThemeDetector;
import io.github.gaming32.superpack.util.GeneralUtilKt;
import io.github.gaming32.superpack.util.HasLogger;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;

import javax.swing.*;
import java.awt.*;
import java.util.function.Consumer;

public final class ProgressDialog extends JDialog implements HasLogger {
    private static final Logger LOGGER = GeneralUtilKt.getLogger();

    private final Consumer<Boolean> themeListener = isDark -> {
        if (SuperpackSettings.INSTANCE.getTheme().isAffectedBySystem()) {
            SwingUtilities.invokeLater(() -> {
                SuperpackSettings.INSTANCE.getTheme().systemThemeChanged(isDark);
                SwingUtilities.updateComponentTreeUI(this);
            });
        }
    };
    private final OsThemeDetector themeDetector;

    private final JLabel note;
    private final JProgressBar progressBar;
    private boolean openedOnce;

    public ProgressDialog(Frame owner, OsThemeDetector themeDetector, String title, String note) {
        super(owner, title, true);
        this.themeDetector = themeDetector;
        themeDetector.registerListener(themeListener);

        setDefaultCloseOperation(DISPOSE_ON_CLOSE);

        this.note = new JLabel(note);
        this.note.setAlignmentX(JLabel.LEFT_ALIGNMENT);

        progressBar = new JProgressBar();
        progressBar.setStringPainted(true);
        progressBar.setAlignmentX(JLabel.LEFT_ALIGNMENT);

        final JPanel cancelPanel = new JPanel();
        cancelPanel.setLayout(new BorderLayout());
        cancelPanel.setBorder(BorderFactory.createEmptyBorder(0, 5, 5, 5));
        // final JButton cancelButton = new JButton("Cancel");
        final JButton cancelButton = new JButton("Cancel");
        cancelButton.addActionListener(ev -> dispose());
        cancelPanel.add(cancelButton, BorderLayout.EAST);
        cancelPanel.setAlignmentX(JLabel.LEFT_ALIGNMENT);

        final JPanel root = new JPanel();
        root.setLayout(new BoxLayout(root, BoxLayout.Y_AXIS));
        root.setBorder(BorderFactory.createEmptyBorder(5, 5, 5, 5));

        root.add(Box.createVerticalGlue());
        root.add(this.note);
        root.add(Box.createVerticalStrut(5));
        root.add(progressBar);
        root.add(Box.createVerticalStrut(5));
        root.add(cancelPanel);
        root.add(Box.createVerticalGlue());

        setContentPane(root);
        pack();
    }

    @Override
    @NotNull
    public Logger getLogger() {
        return LOGGER;
    }

    @Override
    public void setVisible(boolean value) {
        if (value) {
            openedOnce = true;
            themeDetector.registerListener(themeListener);
        } else {
            themeDetector.removeListener(themeListener);
        }
        super.setVisible(value);
    }

    public JLabel getNote() {
        return note;
    }

    public JProgressBar getProgressBar() {
        return progressBar;
    }

    public void setNote(String note) {
        this.note.setText(note);
        pack();
    }

    public void setProgress(int value) {
        progressBar.setValue(value);
    }

    public void setMaximum(int value) {
        progressBar.setMaximum(value);
    }

    public void setIndeterminate(boolean value) {
        progressBar.setIndeterminate(value);
    }

    public void setString(String s) {
        progressBar.setString(s);
    }

    public boolean cancelled() {
        return openedOnce && !isVisible();
    }
}

File: src/main/kotlin/io/github/gaming32/superpack/ProgressDialog.kt
package io.github.gaming32.superpack

import com.jthemedetecor.OsThemeDetector
import io.github.gaming32.superpack.util.HasLogger
import io.github.gaming32.superpack.util.getLogger
import java.awt.BorderLayout
import java.awt.Frame
import javax.swing.*

class ProgressDialog(
    owner: Frame?, private val themeDetector: OsThemeDetector, title: String?, note: String?
) : JDialog(owner, title, true), HasLogger {
    private val themeListener: (Boolean) -> Unit = { isDark ->
        if (SuperpackSettings.INSTANCE.theme.isAffectedBySystem) {
            SwingUtilities.invokeLater {
                SuperpackSettings.INSTANCE.theme.systemThemeChanged(isDark)
                SwingUtilities.updateComponentTreeUI(this)
            }
        }
    }

    val note: JLabel
    val progressBar: JProgressBar
    private var openedOnce = false

    init {
        themeDetector.registerListener(themeListener)

        defaultCloseOperation = DISPOSE_ON_CLOSE

        this.note = JLabel(note)
        this.note.alignmentX = LEFT_ALIGNMENT

        progressBar = JProgressBar()
        progressBar.isStringPainted = true
        progressBar.alignmentX = LEFT_ALIGNMENT

        val cancelPanel = JPanel()
        cancelPanel.layout = BorderLayout()
        cancelPanel.border = BorderFactory.createEmptyBorder(0, 5, 5, 5)

        val cancelButton = JButton("Cancel")
        cancelButton.addActionListener { dispose() }
        cancelPanel.add(cancelButton, BorderLayout.EAST)
        cancelPanel.alignmentX = LEFT_ALIGNMENT

        val root = JPanel()
        root.layout = BoxLayout(root, BoxLayout.Y_AXIS)
        root.border = BorderFactory.createEmptyBorder(5, 5, 5, 5)

        root.add(Box.createVerticalGlue())
        root.add(this.note)
        root.add(Box.createVerticalStrut(5))
        root.add(progressBar)
        root.add(Box.createVerticalStrut(5))
        root.add(cancelPanel)
        root.add(Box.createVerticalGlue())

        contentPane = root
        pack()
    }

    override val logger get() = LOGGER

    override fun setVisible(value: Boolean) {
        if (value) {
            openedOnce = true
            themeDetector.registerListener(themeListener)
        } else {
            themeDetector.removeListener(themeListener)
        }
        super.setVisible(value)
    }

    fun setNote(note: String?) {
        this.note.text = note
        pack()
    }

    var progress by progressBar::value
    var maximum by progressBar::maximum

    var indeterminate
        get() = progressBar.isIndeterminate
        set(value) { progressBar.isIndeterminate = value }

    var string: String? by progressBar::string

    fun cancelled(): Boolean {
        return openedOnce && !isVisible
    }

    companion object {
        private val LOGGER = getLogger()
    }
}
